src/dbnode/encoding/proto/custom_unmarshal.go

```diff
@@ -0,0 +1,375 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package proto
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"sort"
+
+	"github.com/golang/protobuf/proto"
+	dpb "github.com/golang/protobuf/protoc-gen-go/descriptor"
+	"github.com/jhump/protoreflect/desc"
+	"github.com/jhump/protoreflect/dynamic"
+)
+
+var (
+	// Groups in the Protobuf wire format are deprecated, so simplify the code significantly by
+	// not supporting them.
+	errGroupsAreNotSupported = errors.New("use of groups in proto wire format is not supported")
+	zeroValue                unmarshalValue
+)
+
+type customFieldUnmarshaler interface {
+	sortedCustomFieldValues() sortedCustomFieldValues
+	nonCustomFieldValues() *dynamic.Message
+	numNonCustomValues() int
+	resetAndUnmarshal(schema *desc.MessageDescriptor, buf []byte) error
+}
+
+type customUnmarshaler struct {
+	schema *desc.MessageDescriptor
+
+	decodeBuf *buffer
+
+	customValues sortedCustomFieldValues
+
+	nonCustomValues *dynamic.Message
+	numNonCustom    int
+}
+
+func newCustomFieldUnmarshaler() customFieldUnmarshaler {
+	return &customUnmarshaler{
+		decodeBuf: newCodedBuffer(nil),
+	}
+}
+
+func (u *customUnmarshaler) sortedCustomFieldValues() sortedCustomFieldValues {
+	return u.customValues
+}
+
+func (u *customUnmarshaler) numNonCustomValues() int {
+	return u.numNonCustom
+}
+
+func (u *customUnmarshaler) nonCustomFieldValues() *dynamic.Message {
+	if u.nonCustomValues == nil {
+		u.nonCustomValues = dynamic.NewMessage(u.schema)
+	}
+	return u.nonCustomValues
+}
+
+func (u *customUnmarshaler) unmarshal() error {
+	u.customValues = u.customValues[:0]
+
+	if u.nonCustomValues != nil {
+		u.nonCustomValues.Reset()
+	}
+
+	isSorted := true
+	for !u.decodeBuf.eof() {
+		tagAndWireTypeStartOffset := u.decodeBuf.index
+		fieldNum, wireType, err := u.decodeBuf.decodeTagAndWireType()
+		if err != nil {
+			return err
+		}
+
+		fd := u.schema.FindFieldByNumber(fieldNum)
+		if fd == nil {
+			return fmt.Errorf("encountered unknown field with field number: %d", fieldNum)
```

Comment:
I could still handle it by just deferring to the dynamic protobuf library like I do for complex fields (repeated, maps, nested messages, etc) but I think for now I'd prefer to reject the writes. We can always change this later / make it configurable but it potentially impacts performance so much I'd prefer people start getting write errors because they forgot to upgrade their schema instead of having their nodes suddenly start having really bad compression.

Alternatively I could just ignore unknown fields but that seems like the worst of both worlds

Reply:
sounds good, agree we should start with the stricter mode, the alternative is too much delayed surprise.  Client side configuration should require a schema version, which by itself already encourages good behavior, because if application developer forgets to deploy schema, the write will fail at client side due to unable to marshal any new fields (presumably that's primary reason for schema update), when opening a session, schema version is sent as part of the session context, then we can reject the write on dbnode side if dbnode has not gotten the most up to date schema (from etcd watches). Unless there is some serious network partition between meta and db node, this should be a very rare error.

Reply:
We could eventually try to make DB nodes take part in acknowledging the schema update before enabling it and returning success to the client. (i.e. two phase commit for schema updates)

That way the user experience is pretty reasonable since it wouldn't be successful until all nodes acknowledge and reply success by default.  There could also be a mode to avoid the two phase commit if operator desires, that way you could make schema changes even with nodes that are currently unhealthy.